src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -7164,12 +7164,63 @@ bool canFollowNullableType()
+                        if (this.CurrentToken.Kind == SyntaxKind.IdentifierToken)
+                        {
+                            // Given that we are deciding whether we take `?:` operator or not while looking at state after `?`,
+                            // our focus is on `:` token as well. But `:` have special meaning in context of switch statements
+                            // and expressions (since `:` in switch expressions is recovered to `=>`). So consider the following case:
+                            // `case T? t:`. On its own `T ? t :` can be treated as a conditional expression with condition `T`,
+                            // 'whenTrue' `t` and 'whenFalse' not yet parsed. But it is way better to take `:` as case label end and
+                            // thus parse label condition as a declaration pattern `T? t`
+                            if (isTopLevelSwitchPattern)
+                            {
+                                return true;
+                            }
+
+                            // Special case for `await`: if current token is an `await` identifier and the next token
+                            // can start an expression, prefer parsing conditional expression
+                            // rather than having `await` as a pattern designation
+                            if (this.CurrentToken.ContextualKind == SyntaxKind.AwaitKeyword)
+                            {
+                                using var _ = GetDisposableResetPoint(resetOnDispose: true);
+                                this.EatToken();
+                                return !CanStartExpression();
+                            }
+
+                            var nextTokenKind = PeekToken(1).Kind;
+
+                            // These token either 100% end a pattern or start a new one:
+                            // - Literal starts a new pattern. Can occur in list pattern with missing separation `,`: x is [int[]? arr 5] -- missing `,` after `arr`
+                            // - Predefined type is basically the same case: x is [string[]? slice char ch] -- missing `,` after `slice`
+                            // - `)`, `]` and `}` obviously end a pattern: if (x is int? i), indexable[x is string? s], x is { Prop: Type? typeVar }
+                            // - `{` starts a new pattern. Note, that `[` and `(` are not in the list because they can start an invocation/indexer
+                            //   in cases we want a conditional expression: x is char ? dict[a] : 2, x is byte ? func(x) : null
+                            // - `,` end a pattern in list/property pattern: x is { Prop1: Type1? type, Prop2: Type2 }
+                            // - `;` end a pattern if it finishes an expression statement: var y = x is bool? b;
+                            // - EndOfFileToken is obviously the end of parsing. We are better parsing a pattern rather than an unfinished conditional expression
```

Comment:
i really want to tweak all of this.  can i push to this branch?

Reply:
Sure. The PR has already taken its (most likely) final shape, these are just small things for readability/understandability, so it shouldn't take you long

Reply:
Ok. The pushed code is how i prefer code like thsi to eb written.  Collapsing into one large expression, with dense comments to explain it all is hard to read/maintain to me.  Breaking into clear checks with easy comments on each check makes things much more manageable/maintainable imo.